File: solutions/c/reverse-string/1/reverse_string.c
#include "reverse_string.h"

#include <stddef.h>
#include <stdlib.h>

char *reverse(const char *value) {
    size_t len = 0;
    for (const char *ch = value; *ch != '\0'; ch++) {
        len++;
    }
    char *reversed = calloc(len + 1, sizeof(char));
    char *from_end = reversed + len - 1;
    for (const char *ch = value; *ch != '\0'; ch++) {
        *from_end = *ch;
        from_end--;
    }
    return reversed;
}

File: solutions/zig/reverse-string/1/reverse_string.zig
/// Writes a reversed copy of `s` to `buffer`.
pub fn reverse(buffer: []u8, s: []const u8) []u8 {
    for (s, 0..) |c, i| {
        buffer[s.len - 1 - i] = c;
    }
    return buffer;
}
